Retry on oversize file

bk2204 opened the pull request:
If we're attempting to resume a download but the file we have on disk is the expected size or too large, we'll send an invalid Range header where the lower bound is greater than the upper bound. If this situation occurs, simply retry the download from the beginning, since we clearly don't have the expected data (or we wouldn't be resuming) and it isn't clear how else to recover.

This fixes #3494.

bk2204 commented on t/t-resume-http-range.sh:
```diff
@@ -85,3 +85,42 @@ begin_test "resume-http-range-fallback"
+  printf "%s" "${contents/st/aa}" >".git/lfs/incomplete/$contents_oid.tmp"
+
+  GIT_TRACE=1 git lfs fetch 2>&1 | tee fetchresume.log
+  GIT_TRACE=1 git lfs fetch 2>&1 | tee fetchresume.log
```
Yeah, that's the idea. I probably need to add a comment to that effect here.

I'll add some assertions that we've deleted the problematic file and ensure that the new file has the correct hash.

ttaylorr replied:
I'm not sure that that behavior is worth testing explicitly. I think that what we care about at the integration level is that the fetch was executed successfully (by however that fetch actually works).

Testing an implementation like deleting the file in question makes such a test brittle in my mind.

bk2204 replied:
I've decided to leave it as it is and add some comments to help make the situation a bit easier to understand.

ttaylorr commented on t/cmd/lfstest-gitserver.go:
```diff
@@ -749,6 +749,24 @@ func storageHandler(w http.ResponseWriter, r *http.Request) {
 					byteLimit = 8
 					batchResumeFailFallbackStorageAttempts++
 				}
+			} else if len(by) == len("status-batch-retry") && string(by) == "status-batch-retry" {
```
We may be copying what's already here, but is there a need to check the string both for string and length equality? Shouldn't the later follow from the former?

bk2204 replied:
I think this was an optimization (perform the cheap check first), but I'm happy to just remove it and switch to a string comparison. I expect Go does the same comparison internally anyway.

ttaylorr commented on t/cmd/lfstest-gitserver.go:
```diff
@@ -749,6 +749,24 @@ func storageHandler(w http.ResponseWriter, r *http.Request) {
+					if match != nil && len(match) > 2 && len(match[2]) != 0 {
+						first, _ := strconv.ParseInt(match[1], 10, 32)
+						second, _ := strconv.ParseInt(match[2], 10, 32)
+						if second < first {
```
Do you think it would be worth adding a comment indicating what it means to be in this conditional?

bk2204 replied:
Sure, I can add comments for these conditionals.

ttaylorr commented on t/cmd/lfstest-gitserver.go:
```diff
@@ -749,6 +749,24 @@ func storageHandler(w http.ResponseWriter, r *http.Request) {
+							return
+						}
+					}
+					if match != nil && len(match) > 1 {
```
Likewise in this conditional?

ttaylorr commented on t/t-resume-http-range.sh:
```diff
@@ -85,3 +85,42 @@ begin_test "resume-http-range-fallback"
+
+  # this string announces to server that we want a test that
+  # interrupts the transfer when started from 0 to cause resume
+  contents="status-batch-retry"
```
Thanks for documenting this. When I first explored these tests, I was often confused by why these magic strings caused certain things to happen; this is really helpful for similar newcomers.

bk2204 replied:
I actually need to revise this, since this comment is pulled in from another test assertion. But I will replace it with a hopefully equally helpful comment.

bk2204 commented:
I think I've addressed all of the concerns here. There are now copious correct comments, and I've removed the “optimization” which Go likely already performs.
